home work 3

AlexL-1 opened the pull request:
Домашнее задание №3. Начинающий реакт разработчик. 
Реализована страница с продуктами и страница с одним продуктом.
Есть переключатель страниц. Фильтр и поиск не работают.

(кто-то балуется с API. Раньше там было 84 товара, а теперь только 22)
https://api.escuelajs.co/api/v1/products?offset=0&limit=10

AlexL-1 committed:
home work 3

Iraida-ivanova commented on src/utils/axios_get.ts:
```diff
@@ -0,0 +1,30 @@
+      const response = await axios.get(url, {
+        params: params
+      });
+        return response.data as T;
```
ты тут приводишь к типу T, но запрос может вернуть не такой тип и ты пропустишь ошибку

AlexL-1 replied:
Прочел, что приводить типы не глядя нельзя, может сломаться в run-time. Значит надо будет в другом месте проверить.
https://dev.to/brainiacneit/understanding-typescript-as-keyword-11db

Iraida-ivanova commented on src/utils/axios_get.ts:
```diff
@@ -0,0 +1,30 @@
+
+
+// T неизвестный пока тип. Мы будем давать тип ожидаемого ответа при обращении  к функции
+export const makeGetRequest = async<T> ({url,params }:RequestParams) => {
```
```suggestion
export const makeGetRequest = async<T> ({url,params }:RequestParams): Promise<T> => {
```

Iraida-ivanova replied:
Тебе нужно обработать кейс, когда запрос вернет ошибку и что ты будешь возвращать в этом случае. Можно создать такой тип 
```
export type BaseResponse<T> = {
data: T;
isError: false
} | {isError: true;}

export const makeGetRequest = async<T> ({url,params }:RequestParams): Promise<BaseResponse<T>> => {
```

AlexL-1 replied:
пока такие конструкции с generic для меня трудны. Т.е. когда читаешь код, то понятно, а сам так писать не станешь, а напишешь, как умеешь, через catch. Но все жё реализовал этот  BaseResponse<T>, чтобы остался пример того, как надо такие вещи писать.

Iraida-ivanova commented on src/utils/axios_get.ts:
```diff
@@ -0,0 +1,30 @@
+    try {
+      const response = await axios.get(url, {
+        params: params
+      });
```
здесь обработать
```
if (response.status === 200) {
return {
data: response.data,
isError: false
}
} 
return {
isError: true
}
```
Затем при использовании этой функции ты будешь проверять  if (response.isError)  {...

AlexL-1 replied:
добавил проверку на код 200

Iraida-ivanova commented on src/utils/axios_get.ts:
```diff
@@ -0,0 +1,30 @@
+      //console.log(response);
+    } catch (error:any) {
+        throw new Error(error)
+      //console.error(error);
```
и здесь ты можешь не пробрасывать ошибку, а возвращать return {isError: true}

AlexL-1 replied:
переписал на тип  BaseResponse<T> и теперь в catch возвращаю {isError: true}

Iraida-ivanova commented on src/styles/variables/global.scss:
```diff
@@ -0,0 +1,37 @@
+
+  //shadow for dropdown and card
+  --shadow: 0px 4px 10px 0px #00000040;
+}
```
Можно заменить на scss переменные. И в отдельном файлике написать несколько миксинов

AlexL-1 replied:
заменил на $. Cделал миксины для Loader. А тень у меня только одна для всех. Сделал из неё миксин без входных параметров.
@mixin shadow()
{
shadow: 0px 4px 10px 0px #00000040;
}

Iraida-ivanova commented on src/api/getProduct.ts:
```diff
@@ -0,0 +1,12 @@
+//надо приклетить id к url
+const getProduct = async (id: number ) => {
+  const data = await makeGetRequest<Product>({url:API_ENDPOINTS.PRODUCTS + "/" + id.toString()});
+  return data;
```
вот здесь до return можно сделать проверку и если response.isError возвращать либо тот же BaseResponse(c {isError: true}), либо null

AlexL-1 replied:
добавил проверку

Iraida-ivanova commented on src/api/getProduct.ts:
```diff
@@ -0,0 +1,12 @@
+//https://api.escuelajs.co/api/v1/products/4
+//надо приклетить id к url
+const getProduct = async (id: number ) => {
+  const data = await makeGetRequest<Product>({url:API_ENDPOINTS.PRODUCTS + "/" + id.toString()});
```
```
`${API_ENDPOINTS.PRODUCTS}/${id}`
```

AlexL-1 replied:
да, надо все эти сложения строк переделать на шаблонные строки

Iraida-ivanova commented on src/api/getProduct.ts:
```diff
@@ -0,0 +1,12 @@
+
+//https://api.escuelajs.co/api/v1/products/4
+//надо приклетить id к url
+const getProduct = async (id: number ) => {
```
```suggestion
const getProduct = async (id: number ): Promise<Product | null> => {
```
или
`const getProduct = async (id: number ): Promise<BaseResponse<Product>> => {`

Iraida-ivanova commented on src/api/getAllProducts.ts:
```diff
@@ -0,0 +1,11 @@
+  return data;
+};
+
+export default getAllProducts;
```
Здесь такие же  комментарии, как в getProduct

AlexL-1 replied:
добавил и тут возвращение промиса.

Iraida-ivanova commented on src/api/types.ts:
```diff
@@ -0,0 +1,49 @@
+//на самом деле там есть ещё поля даты создания
+/*
+[{"id":10,"title":"Product 1 Updated","price":1000,"description":"Description 1 Updated","images":["[\"https://i.imgur.com/wXuQ7bm.jpeg\"","\"https://i.imgur.com/BZrIEmb.jpeg\"","\"https://i.imgur.com/KcT6BE0.jpeg\"]"],"creationAt":"2024-04-18T20:15:40.000Z","updatedAt":"2024-04-19T16:39:49.000Z","category":{"id":1,"name":"Clothes","image":"https://i.imgur.com/QkIa5tT.jpeg","creationAt":"2024-04- ...
+*/
```
ненужные комментарии нужно будет удалить

AlexL-1 replied:
Вероятно, что все мои комментарии следует признать ненужными. Просто я, как начинающий, пишу пояснения там, где опытному и так всё понятно. Ещё увидел, что есть специальные коментарии /** */ их оставлю.

Iraida-ivanova commented on src/api/types.ts:
```diff
@@ -0,0 +1,49 @@
+images: string, //string[] does not work here
+creationAt: string, //"2024-04-18T20:15:40.000Z"
+updatedAt: string,//"2024-04-18T20:15:40.000Z"
+}
```
нужно настроить линтер, чтобы было красиво, и хотя бы перед коммитом запускай  yarn run lint:fix

AlexL-1 replied:
что-то Prettier ESLint не срабатывал в VSCode при сохранении файла. Переключился на другой Prettier 
![image](https://github.com/AlexL-1/ecommerce/assets/110930042/c8e1c9b4-3c01-41e3-94db-65863c0c8d51).
Ещё нашел команды, пишу, чтобы не забыть: yarn prettier . --write,    yarn run lint --fix

Iraida-ivanova commented on src/components/Loader/Loader.tsx:
```diff
@@ -0,0 +1,21 @@
+  className?: string;
+};
+
+const Loader: React.FC<LoaderProps> = ({ size, className }: LoaderProps) => {
```
```suggestion
const Loader: React.FC<LoaderProps> = ({ size = "s", className }: LoaderProps) => {
```

Iraida-ivanova commented on src/components/Loader/Loader.tsx:
```diff
@@ -0,0 +1,21 @@
+};
+
+const Loader: React.FC<LoaderProps> = ({ size, className }: LoaderProps) => {
+  if (!size) size = "s";
```
```suggestion
```

AlexL-1 replied:
в другом замечании увидел лучший способ - передавать дефолтное значение. Так и сделаю
({ size="s", className }: LoaderProps)

Iraida-ivanova commented on src/components/Loader/Loader.module.scss:
```diff
@@ -0,0 +1,52 @@
+  border-top-color: transparent;
+  transform: rotate(125deg);
+
+  &.l
```
```suggestion
  &_l
```

AlexL-1 replied:
действительно, если точка(.) то амперсанд не нужен, просто вложить одно в другое.

Iraida-ivanova commented on src/components/Loader/Loader.tsx:
```diff
@@ -0,0 +1,21 @@
+const Loader: React.FC<LoaderProps> = ({ size, className }: LoaderProps) => {
+  if (!size) size = "s";
+
+  let classNameString: string = styles.loader + " " + size;
```
```suggestion
  let classNameString: string = styles.loader + " " + styles.loader_size;
```

AlexL-1 replied:
тут я не понял, как переменная size попадет в loader_size.  Сделал через if else
if (size === "s") classNameString += " " + styles.loader_s;

Iraida-ivanova commented on src/components/Loader/Loader.tsx:
```diff
@@ -0,0 +1,21 @@
+export type LoaderProps = {
+  /** Размер */
+  size?: "s" | "m" | "l";
+  /** Дополнительный класс */
```
Если по неймингу понятно, комментарии не нужны

AlexL-1 replied:
да, тут очевидно всё. убрал

Iraida-ivanova commented on src/components/Loader/Loader.module.scss:
```diff
@@ -0,0 +1,52 @@
+
+
+
+
```
тут тоже должно быть красиво

AlexL-1 replied:
запустил yarn prettier . --write и он поправил и это файл, хоть и пишет "unchanged"
![image](https://github.com/AlexL-1/ecommerce/assets/110930042/3fd8c6eb-8052-485d-8112-c97853262cfb)

Iraida-ivanova commented on src/components/Loader/Loader.module.scss:
```diff
@@ -0,0 +1,52 @@
+    height:32px;
+    border-width: 4px;
+    margin-top: 8px;
+    margin-left: 8px;
```
вот такие повторяющиеся стили можно вынести в миксин в верхней части файла
например, 
```
@mixin loaderSize($size, $borderWidth, $margin) {
width:$size;
    height:$size;
    border-width: $borderWidth;
    margin-top: $margin;
    margin-left: $margin;
}
```
и потом использовать так `@include loaderSize($size: 32px, $borderWidth: 4px, $margin: 8px)`

AlexL-1 replied:
сделал миксин

Iraida-ivanova commented on src/components/Card/Card.tsx:
```diff
@@ -0,0 +1,72 @@
+          {props.subtitle}
+        </Text>
+        <div
+          style={{
```
лучше использовать className

AlexL-1 replied:
убрал все style={{}} блоки и дал className

Iraida-ivanova commented on src/components/Input/Input.tsx:
```diff
@@ -0,0 +1,44 @@
+const Input = React.forwardRef<HTMLInputElement, InputProps>(
+  ({ value, onChange, afterSlot, className, ...props }: InputProps, ref) => {
+    let classNameString: string = styles.input;
+    if (afterSlot) classNameString += " " + styles.inputLoading;
```
установи [classnames](https://classic.yarnpkg.com/en/package/classnames), будет удобнее работать с классами

AlexL-1 replied:
установил, посмотрел примеры использования. Буду писать так:
const classNames = require('classnames');
classNames('foo', 'bar'); // => 'foo bar'

Iraida-ivanova commented on src/components/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,90 @@
+    return () => {
+      document.removeEventListener("mousedown", handleOutsideClick);
+    };
+  });
```
```suggestion
  }, []);
```
Так думаю, нужно навесить обработчик при монтировании. Без массива зависимостей этот эффект будет выззываться при каждом перерендере

Iraida-ivanova commented on src/components/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,90 @@
+  onChange,
+}: DropdownProps) => {
+  let [isExpanded, Toggle] = useState(false);
+  let [index, SetIndex] = useState(keySelected);
```
```suggestion
  let [index, setIndex] = useState<string>(keySelected);
```

Iraida-ivanova commented on src/components/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,90 @@
+  keySelected,
+  onChange,
+}: DropdownProps) => {
+  let [isExpanded, Toggle] = useState(false);
```
```suggestion
  let [isExpanded, setIsExpanded] = useState<boolean>(false);
```
так понятнее, что это функция, изменяющая state

Iraida-ivanova commented on src/components/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,90 @@
+    }
+  };
+
+  const handleOnListClicked = (key: string) => {
```
потом желательно такие функции в useCallback завернуть

AlexL-1 replied:
таки да, пересмотрел ещё раз видео и понял, что без useCallback будет срабатывать рендеринг лишний раз

Iraida-ivanova commented on src/components/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,90 @@
+      //only 1 category
+      // if selected, then unselect
+      if (key == index) {
+        onChange(key); //callback
```
если тот же key, чаще всего не нужно вызывать onChange, тем более с тем же key. Раз ты сбрасываешь key, то и onChange должен вызываться с "".

AlexL-1 replied:
пожалуй, переключение страниц оставлю на домашнее задание 4. И все замечания там уже реализую.

Iraida-ivanova commented on src/components/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,90 @@
+      onClick={() => {
+        handleOnListClicked(elm.key);
+      }}
+      className={elm.key == index ? styles.itemSelected : ""}
```
```suggestion
      className={elm.key === index ? styles.itemSelected : ""}
```
Желательно везде использовать оператор строгого равенства

AlexL-1 replied:
поправил везде, где нашел.

Iraida-ivanova commented on src/components/Navbar/Navbar.tsx:
```diff
@@ -0,0 +1,49 @@
+  return (
+    <div className={styles.Navbar}>
+      <div className={styles.left}>
+        <Link to="/">
```
Здесь можно использовать `<NavLink `

AlexL-1 replied:
прочел, что NavLink это специальный Link для навигации. Сделаю NavLink в меню.
https://reactrouter.com/en/main/components/nav-link

Iraida-ivanova commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,114 @@
+    //first and last pages are already added
+    for (let i = 2; i < totalPages - 1; i++) {
+      if (currentPage == i) {
+        pages.push(<a className={styles.currentPage}>{i}</a>);
```
Pages нужно сделать state.  И в pages могут быть просто номера страниц, для которых можно нарисовать  отдельный компонент(или элемент), который ,будет использовать onClick, переданный сверху. Компонент Pagination не должен рулить какой-то логикой. Он должен уметь подсвечивать выбранную страницу, использовать onClick и отображаться в соответствии с номером, выбранной  страницы.

AlexL-1 replied:
пожалуй, переключение страниц оставлю на домашнее задание 4. И все замечания там уже реализую

Iraida-ivanova commented on src/pages/ProductDetail/ProductDetail.tsx:
```diff
@@ -0,0 +1,95 @@
+import { useNavigate } from "react-router-dom";
+
+const ProductDetail = () => {
+  let id: number; //id of the element to show.
```
можно `let { id } = useParams();`

AlexL-1 replied:
переделал: вместо product?id=1 теперь product/1 
и использую useParams();

Iraida-ivanova commented on src/pages/ProductDetail/ProductDetail.tsx:
```diff
@@ -0,0 +1,95 @@
+    id = 1; //at least we will show some product.
+  }
+
+  const [productData, SetProductData] = useState<Product>();
```
```suggestion
  const [productData, SetProductData] = useState<Product | null>(null);
```

Iraida-ivanova commented on src/pages/ProductDetail/ProductDetail.tsx:
```diff
@@ -0,0 +1,95 @@
+        Back
+      </a>
+
+      <div className={styles.cardInfo}>
```
{
```suggestion
      {productData && (<div className={styles.cardInfo}>
```

Iraida-ivanova reviewed (commented):
1)App.tsx и pages лучше вынести в одну папку App, так как в App.tsx ты определяешь paths и страницы для них(они лежат в pages)
2) После изучения новых хуков желательно их применить там, где они нужны
3) комментарии желательно применять для всех похожих мест

AlexL-1 committed:
Update src/utils/axios_get.ts

function type added

Co-authored-by: Iraida Ivanova <81653058+Iraida-ivanova@users.noreply.github.com>

AlexL-1 committed:
Update src/api/getProduct.ts

указать возвращаемый тип

Co-authored-by: Iraida Ivanova <81653058+Iraida-ivanova@users.noreply.github.com>

AlexL-1 committed:
Update src/pages/ProductDetail/ProductDetail.tsx

в useState писать null, ведь ещё не загрузили продукты

Co-authored-by: Iraida Ivanova <81653058+Iraida-ivanova@users.noreply.github.com>

AlexL-1 committed:
Update src/pages/ProductDetail/ProductDetail.tsx

условный рендеринг. Показыать карточку, только если есть данные.

Co-authored-by: Iraida Ivanova <81653058+Iraida-ivanova@users.noreply.github.com>

AlexL-1 committed:
Update src/components/Loader/Loader.tsx

добавить дефолтное значение для размера

Co-authored-by: Iraida Ivanova <81653058+Iraida-ivanova@users.noreply.github.com>

AlexL-1 committed:
Update src/components/Dropdown/Dropdown.tsx

пропустил массив зависимостей []. Без него useEffect срабатывает при каждом перерендере, а не только один раз вначале

Co-authored-by: Iraida Ivanova <81653058+Iraida-ivanova@users.noreply.github.com>

AlexL-1 committed:
Update src/components/Dropdown/Dropdown.tsx

useState это generic, надо указывать тип.

Co-authored-by: Iraida Ivanova <81653058+Iraida-ivanova@users.noreply.github.com>

AlexL-1 committed:
Update src/components/Dropdown/Dropdown.tsx

в useState принято называть функцию со словом set.
надо в generic писать тип.

Co-authored-by: Iraida Ivanova <81653058+Iraida-ivanova@users.noreply.github.com>
